src/borrow_tracker/mod.rs

```diff
@@ -0,0 +1,340 @@
+    Off,
+}
+
+impl Method {
+    pub fn is_active(self) -> bool {
+        !matches!(self, Method::Off)
+    }
+
+    pub fn instanciate_global_state(self, config: &MiriConfig) -> Option<GlobalState> {
+        self.is_active().then(|| {
+            RefCell::new(GlobalStateInner::new(
+                config.tracked_pointer_tags.clone(),
+                config.tracked_call_ids.clone(),
+                config.retag_fields,
+            ))
+        })
+    }
+
+    pub fn new_allocation<'mir, 'tcx>(
+        self,
+        id: AllocId,
+        alloc: &Allocation,
+        global: Option<&GlobalState>,
+        kind: MemoryKind<machine::MiriMemoryKind>,
+        machine: &MiriMachine<'mir, 'tcx>,
+    ) -> Option<State> {
+        match self {
+            Method::StackedBorrows =>
+                Some(State::Stack(RefCell::new(Stacks::new_allocation(
+                    id,
+                    alloc.size(),
+                    global.unwrap(),
+                    kind,
+                    machine,
+                )))),
+            Method::Off => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum State {
+    Stack(stacked::AllocExtra),
+}
+
+impl State {
+    pub fn assume_sb(&self) -> &stacked::AllocExtra {
+        match self {
+            State::Stack(ref sb) => sb,
+        }
+    }
+
+    pub fn before_memory_read<'tcx, 'mir>(
+        &self,
+        alloc_id: AllocId,
+        prov_extra: ProvenanceExtra,
+        range: AllocRange,
+        state: &GlobalState,
+        machine: &MiriMachine<'mir, 'tcx>,
+    ) -> InterpResult<'tcx> {
+        match self {
+            State::Stack(sb) =>
+                sb.borrow_mut().before_memory_read(alloc_id, prov_extra, range, state, machine),
+        }
+    }
+
+    pub fn before_memory_write<'tcx, 'mir>(
+        &mut self,
+        alloc_id: AllocId,
+        prov_extra: ProvenanceExtra,
+        range: AllocRange,
+        state: &GlobalState,
+        machine: &MiriMachine<'mir, 'tcx>,
+    ) -> InterpResult<'tcx> {
+        match self {
+            State::Stack(sb) =>
+                sb.get_mut().before_memory_write(alloc_id, prov_extra, range, state, machine),
+        }
+    }
+
+    pub fn before_memory_deallocation<'tcx, 'mir>(
+        &mut self,
+        alloc_id: AllocId,
+        prov_extra: ProvenanceExtra,
+        range: AllocRange,
+        state: &GlobalState,
+        machine: &MiriMachine<'mir, 'tcx>,
+    ) -> InterpResult<'tcx> {
+        match self {
+            State::Stack(sb) =>
+                sb.get_mut().before_memory_deallocation(alloc_id, prov_extra, range, state, machine),
+        }
+    }
+
+    pub fn remove_unreachable_tags(&self, tags: &FxHashSet<BorTag>) {
+        match self {
+            State::Stack(sb) => sb.borrow_mut().remove_unreachable_tags(tags),
+        }
+    }
+
+    pub fn expose_tag(&self, tag: BorTag) {
+        match self {
+            State::Stack(sb) => sb.borrow_mut().expose_tag(tag),
+        }
+    }
```

Comment:
The fact that these take `&self` is surprising. I think it'll make more sense now or later to hoist the `RefCell` out to `State` instead of sinking it into both the borrow tracking implementations.

Reply:
We'll want the `RefCell` to be inside the `Option` though, so if we add a `None` variant to `State` then we wouldn't want a `RefCell` around that.

Reply:
Actually this could just take `&mut self` I think, the only caller has a mutable reference available.

But I agree that since this file here is doing the `borrow_mut` calls, this file should also be adding the `RefCell`. We are already assuming that `stacked::AllocExtra` is a `RefCell`. So might as well define this more like
```rust
pub enum AllocState {
    Stack(RefCell<stacked::AllocState>),
    None,
}
```

Reply:
Ok I'll do that